unic/char/range/src/range.rs

```diff
@@ -0,0 +1,172 @@
+use std::char;
+use std::collections::Bound;
+use CharIter;
+
+/// A range of unicode code points.
+///
+/// The most idiomatic way to construct this range is through the use of the `chars!` macro:
+///
+/// ```
+/// # #[macro_use] extern crate unic_char_range;
+/// # use unic_char_range::*;
+/// # fn main() {
+/// assert_eq!(chars!('a'..='z'), CharRange::closed('a', 'z'));
+/// assert_eq!(chars!('a'..'z'), CharRange::open_right('a', 'z'));
+/// assert_eq!(chars!(..), CharRange::all());
+/// # }
+/// ```
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
```

Comment:
I think derived `PartialEq` doesn't work for us here, because any case with `high < low` is considered empty for us, but `eq()` would return false depending on the numbers.

For rustc `RangeInclusive`, there's been a talk about having an `enum` for the type with an `Empty` variant. I don't think we want that here, but need to well-define emptiness.

Also a reminder that `CharRange` also needs an `is_empty()`, to not enforce `.len() == 0` to call-sites.

Another matter with empty case is the open question of if either of `low` and `high` matter in equality. IMHO, it doesn't matter, and we can have this:
```
fn eq(&self, other: Self) -> bool {
  if self.is_empty() {
    other.is_empty()
  } else {
    self.low == other.low && self.high == other.high
  }
}
```

What do you think?